Allow the creation of String arrays from an interator of &Option<&str>

**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I am making constant arrays for tests from constant slices

With 1.53 Rust introduces the ability to turn constant slices into iterators. For example. However, the collect for

This does not work:
```rust
        let string_array: ArrayRef = Arc::new([Some("foo"), None, Some("bar")].into_iter().collect::<StringArray>());
```

It produces this error:
```
error[E0277]: a value of type `GenericStringArray<i32>` cannot be built from an iterator over elements of type `&Option<&str>`
  --> arrow_util/src/display.rs:43:92
   |
43 |         let string_array: ArrayRef = Arc::new([Some("foo"), None, Some("bar")].into_iter().collect::<StringArray>());
   |                                                                                            ^^^^^^^ value of type `GenericStringArray<i32>` cannot be built from `std::iter::Iterator<Item=&Option<&str>>`
   |
   = help: the trait `FromIterator<&Option<&str>>` is not implemented for `GenericStringArray<i32>`
```

But this does (for primitive arrays):
```rust
        let int64_array: ArrayRef = Arc::new([Some(-1), None, Some(2)].into_iter().collect::<Int64Array>());
```

Instead, for strings we have to do
```rust
        let string_array: ArrayRef = Arc::new(vec![Some("foo"), None, Some("bar")].into_iter().collect::<StringArray>());
```


**Describe the solution you'd like**
I would like to be able to do (note the lack of `vec!`):

```rust
        let string_array: ArrayRef = Arc::new([Some("foo"), None, Some("bar")].into_iter().collect::<StringArray>());
```

Perhaps by making the `FromIter` implementation fancier (right now it hard codes a `Option<&str>` but the iterator above tives a `&Option<&str>`:  https://github.com/apache/arrow-rs/blob/master/arrow/src/array/array_string.rs#L255-L262

**Additional context**

Note -- the same sort if issue exists for `DictionaryArray` as well

Does not work:
```rust
let dict_array = [Some("a"), None, Some("b")]
                .into_iter().collect::<DictionaryArray::<Int32Type>>();
```
Does work
```rust
let dict_array = vec![Some("a"), None, Some("b")]
                .into_iter().collect::<DictionaryArray::<Int32Type>>();
```

--- a/arrow/src/array/array_string.rs
+++ b/arrow/src/array/array_string.rs
@@ -196,11 +196,26 @@ impl<OffsetSize: StringOffsetSizeTrait> GenericStringArray<OffsetSize> {
     }
 }
 
+impl<'a, Ptr, OffsetSize: StringOffsetSizeTrait> FromIterator<&'a Option<Ptr>>
+    for GenericStringArray<OffsetSize>
+where
+    Ptr: AsRef<str> + 'a,
+{
+    /// Creates a `GenericStringArray` based on an iterator of `Option` references.
+    fn from_iter<I: IntoIterator<Item = &'a Option<Ptr>>>(iter: I) -> Self {
+        // Convert each owned Ptr into &str and wrap in an owned `Option`
+        let iter = iter.into_iter().map(|o| o.as_ref().map(|p| p.as_ref()));
+        // Build a `GenericStringArray` with the resulting iterator
+        iter.collect::<GenericStringArray<OffsetSize>>()
+    }
+}
+
 impl<'a, Ptr, OffsetSize: StringOffsetSizeTrait> FromIterator<Option<Ptr>>
     for GenericStringArray<OffsetSize>
 where
     Ptr: AsRef<str>,
 {
+    /// Creates a `GenericStringArray` based on an iterator of `Option`s
     fn from_iter<I: IntoIterator<Item = Option<Ptr>>>(iter: I) -> Self {
         let iter = iter.into_iter();
         let (_, data_len) = iter.size_hint();
@@ -358,6 +373,7 @@ impl<T: StringOffsetSizeTrait> From<GenericListArray<T>> for GenericStringArray<
 
 #[cfg(test)]
 mod tests {
+
     use crate::array::{ListBuilder, StringBuilder};
 
     use super::*;
@@ -484,17 +500,23 @@ mod tests {
 
     #[test]
     fn test_string_array_from_iter() {
-        let data = vec![Some("hello"), None, Some("arrow")];
+        let data = [Some("hello"), None, Some("arrow")];
+        let data_vec = data.to_vec();
         // from Vec<Option<&str>>
-        let array1 = StringArray::from(data.clone());
+        let array1 = StringArray::from(data_vec.clone());
         // from Iterator<Option<&str>>
-        let array2: StringArray = data.clone().into_iter().collect();
+        let array2: StringArray = data_vec.clone().into_iter().collect();
         // from Iterator<Option<String>>
-        let array3: StringArray =
-            data.into_iter().map(|x| x.map(|s| s.to_string())).collect();
+        let array3: StringArray = data_vec
+            .into_iter()
+            .map(|x| x.map(|s| s.to_string()))
+            .collect();
+        // from Iterator<&Option<&str>>
+        let array4: StringArray = data.iter().collect::<StringArray>();
 
         assert_eq!(array1, array2);
         assert_eq!(array2, array3);
+        assert_eq!(array3, array4);
     }
 
     #[test]

--- a/arrow/src/array/array_string.rs
+++ b/arrow/src/array/array_string.rs
@@ -201,7 +201,7 @@ impl<'a, Ptr, OffsetSize: StringOffsetSizeTrait> FromIterator<&'a Option<Ptr>>
 where
     Ptr: AsRef<str> + 'a,
 {
-    /// Creates a `GenericStringArray` based on an iterator of `Option` references.
+    /// Creates a [`GenericStringArray`] based on an iterator of `Option` references.
     fn from_iter<I: IntoIterator<Item = &'a Option<Ptr>>>(iter: I) -> Self {
         // Convert each owned Ptr into &str and wrap in an owned `Option`
         let iter = iter.into_iter().map(|o| o.as_ref().map(|p| p.as_ref()));
@@ -215,7 +215,7 @@ impl<'a, Ptr, OffsetSize: StringOffsetSizeTrait> FromIterator<Option<Ptr>>
 where
     Ptr: AsRef<str>,
 {
-    /// Creates a `GenericStringArray` based on an iterator of `Option`s
+    /// Creates a [`GenericStringArray`] based on an iterator of `Option`s
     fn from_iter<I: IntoIterator<Item = Option<Ptr>>>(iter: I) -> Self {
         let iter = iter.into_iter();
         let (_, data_len) = iter.size_hint();

